src/sdf/value.cpp

```diff
@@ -0,0 +1,240 @@
+                                                Invalid()));
+}
+
+Value Value::exp() const {
+  return Value(ValueKind::OPERATION,
+               std::make_shared<ValueOperation>(OpCode::EXP, *this, Invalid(),
+                                                Invalid()));
+}
+
+Value Value::log() const {
+  return Value(ValueKind::OPERATION,
+               std::make_shared<ValueOperation>(OpCode::LOG, *this, Invalid(),
+                                                Invalid()));
+}
+
+Value Value::sqrt() const {
+  return Value(ValueKind::OPERATION,
+               std::make_shared<ValueOperation>(OpCode::SQRT, *this, Invalid(),
+                                                Invalid()));
+}
+
+Value Value::floor() const {
+  return Value(ValueKind::OPERATION,
+               std::make_shared<ValueOperation>(OpCode::FLOOR, *this, Invalid(),
+                                                Invalid()));
+}
+
+Value Value::ceil() const {
+  return Value(ValueKind::OPERATION,
+               std::make_shared<ValueOperation>(OpCode::CEIL, *this, Invalid(),
+                                                Invalid()));
+}
+
+Value Value::round() const {
+  return Value(ValueKind::OPERATION,
+               std::make_shared<ValueOperation>(OpCode::ROUND, *this, Invalid(),
+                                                Invalid()));
+}
+
+Value Value::sin() const {
+  return Value(ValueKind::OPERATION,
+               std::make_shared<ValueOperation>(OpCode::SIN, *this, Invalid(),
+                                                Invalid()));
+}
+
+Value Value::cos() const {
+  return Value(ValueKind::OPERATION,
+               std::make_shared<ValueOperation>(OpCode::COS, *this, Invalid(),
+                                                Invalid()));
+}
+
+Value Value::tan() const {
+  return Value(ValueKind::OPERATION,
+               std::make_shared<ValueOperation>(OpCode::TAN, *this, Invalid(),
+                                                Invalid()));
+}
+
+Value Value::asin() const {
+  return Value(ValueKind::OPERATION,
+               std::make_shared<ValueOperation>(OpCode::ASIN, *this, Invalid(),
+                                                Invalid()));
+}
+
+Value Value::acos() const {
+  return Value(ValueKind::OPERATION,
+               std::make_shared<ValueOperation>(OpCode::ACOS, *this, Invalid(),
+                                                Invalid()));
+}
+
+Value Value::atan() const {
+  return Value(ValueKind::OPERATION,
+               std::make_shared<ValueOperation>(OpCode::ATAN, *this, Invalid(),
+                                                Invalid()));
+}
+
+std::pair<std::vector<uint8_t>, std::vector<double>> Value::genTape() const {
+  using VO = std::shared_ptr<ValueOperation>;
+  Context ctx;
+  std::unordered_map<VO, Operand> cache;
+  std::vector<VO> stack;
+  if (kind == ValueKind::OPERATION) stack.push_back(std::get<VO>(v));
+
+  auto getOperand = [&](Value x, std::function<void(Value)> f) {
```

Comment:
Simple recursive algorithm to convert the value AST to IR in `context.h`. `context.h` handles the IR, performs optimization and eventually output a bytecode tape.

We don't do any optimization here, so it is quite simple (except using an explicit loop instead of recursion).